Wait for requests to complete on the server in Session#reset!

jhawthorn opened the pull request:
It's possible when using a JS driver for a request to continue being processed on the server after calling `Capybara.reset_sessions!`. Though the browser requests are stopped, any requests the server had already started processing will continue.

This causes issues when using capybara with DatabaseCleaner, as the the cleaner might run at the same time as the request, which will likely cause the request to error.

This PR adds a thread-safe counter to Capybara::Server which keeps track of the number of requests currently being processed by `Capybara::Server::Middleware`. It exposes public methods `Server#pending_requests?`, which is truthy when there are requests still being processed, and `Server#wait_for_pending_requests`, which will wait up to 60 seconds for requests to finish.

twalpole commented:
Why are you attempting to join the server thread while waiting for the requests to end?  Joining would only succeed if the server thread was ending.  I also need to think about how/if this relates to testing with multiple sessions.

jhawthorn commented:
I used `join` just because it's what `boot` was using. I will change it to a sleep.

twalpole commented:
This is definitely still on the table and I'm considering it for 2.6 -- I haven't spent enough time to think through whether it has any issues with the fact that multiple sessions often share the same Capybara::Server and Middleware

jhawthorn commented:
@twalpole thanks for the update. Is there anything I could do to help move this forward?

twalpole commented on lib/capybara/server.rb:
```diff
@@ -4,22 +4,43 @@
 module Capybara
   class Server
+    class Counter
+      attr_reader :value
```
We don't really need the value we just need a zero? method

twalpole commented on lib/capybara/server.rb:
```diff
@@ -4,22 +4,43 @@
       attr_accessor :error
 
-      def initialize(app)
+      def initialize(app, counter=Counter.new)
```
do we really need to accept a counter here? since it's always passed a new counter in usage couldnt we just create a new counter inside initialize

twalpole commented on lib/capybara/server.rb:
```diff
@@ -35,7 +56,8 @@ def ports
     def initialize(app, port=Capybara.server_port, host=Capybara.server_host)
       @app = app
-      @middleware = Middleware.new(@app)
+      @counter = Counter.new
```
see above

twalpole commented on lib/capybara/server.rb:
```diff
@@ -62,6 +84,16 @@ def responsive?
     end
 
+    def pending_requests?
+      @counter.value > 0
```
see comment about value in Counter class

twalpole commented on lib/capybara/session.rb:
```diff
@@ -111,6 +112,14 @@ def reset!
     #
+    # Wait for any pending server requests to finish
+    #
+    def wait_for_server_requests
```
Does this need to part of the public API?

twalpole commented:
@jhawthorn ok -- I've added a few comments to the current code, but there is a larger concern.  I believe this suffers from the same problem as Issue #1540 .  Namely that the methods on session don't really do what one would think when multiple sessions using the same app are created.  This may or may not be of concern I haven't really decided if it's actually a real problem or just an issue that occurs in artificial conditions.  I think it would actually be more of an issue for this situation than the #1540 though.  I don't think we want to be duplicating the app object for each session, but a quick thought at a fix would be to have each session create a new server thread with it's own middleware instance on it's own port but share the app object, which would allow tracking requests on a per session basis.  That could lead to threading problems though and probably requires more thought

twalpole commented:
The per_session_server branch may be required to have this code work with multiple sessions correctly, more comments about it over on Issue #1540

jhawthorn commented:
@twalpole sorry for the delay.

I've addressed your feedback and rebased against the latest master.

As far as I can tell this works fine with the changes in the `per_session_server` branch (different servers would count their requests separately). The code is very similar to that which handles errors, so I would expect them to behave largely the same.

twalpole commented:
Closing this PR as it was merged into PR #1637
